crates/pipeline-manager/src/compiler/util.rs

```diff
@@ -137,6 +238,103 @@ pub(crate) async fn list_content(
     Ok(content)
 }
 
+/// Lists all the named content of a directory.
+/// Named content refers to the files and directories that have a file name.
+/// If a directory entry is detected which is not a file or directory,
+/// is a symlink, or which does not have a file name, an error is thrown.
+async fn list_named_content(dir: &Path) -> Result<Vec<(String, bool)>, UtilError> {
+    let mut content = vec![];
+    for (path, name) in list_content(dir).await? {
+        let Some(name) = name else {
+            return Err(UtilError::InvalidDirEntry(format!(
+                "Directory entry does not have name: '{}'",
+                path.display()
+            )));
+        };
+        if path.is_symlink() {
+            return Err(UtilError::InvalidDirEntry(format!(
+                "Directory entry is unexpectedly a symlink: '{}'",
+                path.display()
+            )));
+        } else if path.is_file() {
+            content.push((name, true))
+        } else if path.is_dir() {
+            content.push((name, false))
+        } else {
+            return Err(UtilError::InvalidDirEntry(format!(
+                "Directory entry is not a file nor directory: '{}'",
+                path.display()
+            )));
+        }
+    }
+    content.sort();
+    Ok(content)
+}
+
+/// Validates the named content of a directory with an expectation.
+/// The expectation is not met if either a file/directory that should
+/// be there is not, or if a file/directory that shouldn't be there is.
+/// An error is returned if the expectation is not met.
+pub(crate) async fn validate_content(
+    dir: &Path,
+    expected: &Vec<(&str, bool)>,
+) -> Result<(), UtilError> {
+    let content = list_named_content(dir).await?;
+    for (name, is_file) in expected {
+        if !content.contains(&(name.to_string(), *is_file)) {
+            return Err(UtilError::InvalidContent(format!(
+                "{} '{}' is missing in: '{}'",
+                if *is_file { "File" } else { "Directory" },
+                name,
+                dir.display()
+            )));
+        }
+    }
+    for (name, is_file) in content {
```

Comment:
this is quadratic in the size of the vector. Hopefully it won't matter.
How about creating a struct (DirectoryContents) for the type of expected?
then many of these functions become methods of that struct

Reply:
It shouldn't matter, as the sizes are currently relatively. It can of course be optimized further.

Hmm, that could work with regards to making it a struct, will try it out

Reply:
Using struct now, the function is still quadratic in complexity.